libraries-5/src/main/java/com/baeldung/yauaa/UserAgentAttributeLoggingFilter.java

```diff
@@ -0,0 +1,45 @@
+package com.baeldung.yauaa;
+
+import java.io.IOException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import nl.basjes.parse.useragent.UserAgent;
+import nl.basjes.parse.useragent.UserAgentAnalyzer;
+
+@Component
+public class UserAgentAttributeLoggingFilter extends OncePerRequestFilter {
+
+    private static final Logger log = LoggerFactory.getLogger(UserAgentAttributeLoggingFilter.class);
+    private final UserAgentAnalyzer userAgentAnalyzer;
+
+    public UserAgentAttributeLoggingFilter(UserAgentAnalyzer userAgentAnalyzer) {
+        this.userAgentAnalyzer = userAgentAnalyzer;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+        throws ServletException, IOException {
+        String userAgentString = request.getHeader("User-Agent");
+
+        if (StringUtils.hasText(userAgentString)) {
+            UserAgent userAgent = userAgentAnalyzer.parse(userAgentString);
+            userAgent
+                .getAvailableFieldNamesSorted()
```

Comment:
How does this differ from getCleanedAvailableFieldNamesSorted?

Reply:
The getCleanedAvailableFieldNamesSorted() method does not include the fields where the value is unavailable. In the article, since we show the unavailable value fields, it'll be better to use getAvailableFieldNamesSorted(). But I've added in a new sentence to mention the other method as well.